File: benchmarks/tsb/bench_combine_first_series.ts
/**
 * Benchmark: combineFirstSeries (standalone) — fill missing values from another Series.
 * Outputs JSON: {"function": "combine_first_series", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { combineFirstSeries, Series } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const data1: (number | null)[] = Array.from({ length: SIZE }, (_, i) =>
  i % 3 === 0 ? null : i * 0.5,
);
const data2 = Array.from({ length: SIZE }, (_, i) => i * 0.1);
const s1 = new Series({ data: data1 });
const s2 = new Series({ data: data2 });

for (let i = 0; i < WARMUP; i++) {
  combineFirstSeries(s1, s2);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  combineFirstSeries(s1, s2);
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "combine_first_series",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_combine_first_series.py
"""Benchmark: Series.combine_first (standalone equivalent) — fill missing values from another Series.
Mirrors tsb bench_combine_first_series.ts for pandas.
"""
import json, time
import pandas as pd
import numpy as np

SIZE = 10_000
WARMUP = 5
ITERATIONS = 50

data1 = [None if i % 3 == 0 else i * 0.5 for i in range(SIZE)]
data2 = [i * 0.1 for i in range(SIZE)]
s1 = pd.Series(data1)
s2 = pd.Series(data2)

for _ in range(WARMUP):
    s1.combine_first(s2)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s1.combine_first(s2)
    times.append((time.perf_counter() - t0) * 1000)

total = sum(times)
mean = total / ITERATIONS
print(json.dumps({
    "function": "combine_first_series",
    "mean_ms": round(mean, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
